sdk/src/main/kotlin/io/provenance/scope/sdk/mailbox/PollAffiliateMailbox.kt

```diff
@@ -0,0 +1,92 @@
+import io.provenance.scope.toHex
+import org.slf4j.LoggerFactory
+import java.util.UUID
+
+typealias MailHandlerFn = (MailboxEvent) -> Boolean?
+
+class PollAffiliateMailbox(val osClient: OsClient, val signingKeyRef: KeyRef, val encryptionKeyRef: KeyRef, val maxResults: Int, val mainNet: Boolean, val handler: MailHandlerFn): Runnable {
+    private val log = LoggerFactory.getLogger(this::class.java)
+
+    override fun run() {
+        osClient.mailboxGet(encryptionKeyRef.publicKey, maxResults).forEach { (mailUuid, dimeInputStream) ->
+            dimeInputStream.getDecryptedPayload(encryptionKeyRef).use {
+                val bytes = it.readAllBytes()
+
+                if (!it.verify()) {
+                    log.warn("Mailbox object verification failure [public key: ${encryptionKeyRef.publicKey.toHex()}] [mailbox uuid: $mailUuid]")
+                    return
+                }
+
+                log.trace("Received mail from poll $mailUuid")
+
+                if (!dimeInputStream.metadata.containsKey(MailboxMeta.KEY)) {
+                    osClient.mailboxAck(mailUuid).also { log.warn("Unhandled mailbox meta: {}", dimeInputStream.metadata) }
+                    return
+                }
+
+                val mailboxKey = dimeInputStream.metadata[MailboxMeta.KEY]
+
+                when (mailboxKey) {
+                    MailboxMeta.FRAGMENT_REQUEST, MailboxMeta.FRAGMENT_RESPONSE -> handleEnvelope(mailUuid, mailboxKey, dimeInputStream.dime.owner, bytes)
+                    MailboxMeta.ERROR_RESPONSE -> handleEnvelopeError(mailUuid, mailboxKey, dimeInputStream.dime.owner, bytes)
+                    else -> throw IllegalStateException("Unhandled mailbox key: $mailboxKey uuid: $mailUuid")
+                }
+            }
+        }
+    }
+
+    private fun handleEnvelope(mailUuid: UUID, mailboxKey: String, ownerAudience: Audience, message: ByteArray) {
+        val envelope = Envelope.parseFrom(message)
+        // todo: packing scope as an any so we don't have to pull in the raw protos just for this... stupid or not?
+        val scope = envelope.scopeOrNull()
+
+        // todo: validation on various uuids being present? Do we still have an execution uuid in play?
+
+        val className = envelope.contract.definition.resourceLocation.classname
+
+        val signingAddress = signingKeyRef.publicKey.getAddress(mainNet)
+
+        // verify affiliate is present on contract/scope
+        envelope.contract.recitalsList.map { it.signer.signingPublicKey.toPublicKey().getAddress(mainNet) }.plus(scope?.ownersList?.map { it.address } ?: listOf())
+            .firstOrNull { it == signingAddress }
+            .orThrow { IllegalStateException("Can't find party on contract execution ${envelope.executionUuid.value} with key ${signingKeyRef.publicKey.toHex()}") }
+
+        // todo: do we need to provide any sort of contract whitelisting mechanism, or just let the client ignore anything they don't want?
```

Comment:
I think this can be ignored. Due to the way os-locator and object-store replication is currently configured, you can't just mail random stuff to a public key and have them need to handle it. I guess helping with that might have been the original intent of this?